ls1intum/Artemis
# `Programming exercises`: Enable recreation of build plans in Jenkins

### Checklist
#### General


#### Server
#### Changes affecting Programming Exercises

### Motivation and Context
Artemis should be able to recreate the solution and template build plans on Jenkins if they get lost there, as they don’t contain any state. This feature already was implemented for Bamboo, but not for Jenkins. This PR enables the feature for Jenkins as well.

### Description
The interface for the continuous integration service already provided the functionality to recreate the build plans. For Jenkins however, the implementation was empty.

User-facing difference: The button ‘Recreate Build Plans’ in the exercise details page for the instructor now works on Jenkins.

### Steps for Testing
Prerequisites:
- 1 Instructor
- 1 Programming Exercise
- Testing should be primarily on TS2. as this PR changes Jenkins interactions.

1. Go to the exercise overview page as the instructor.
2. The ‘Check Integrity’ button should report that no issues are found.
3. At the bottom of the page are links to the build plans in Jenkins. Logging into Jenkins with the username and password combination of `kl****` you can find in Confluence should work.
4. Repeat with a few {variations}:
    - Participate in/or resume the exercise as a student. You can use the same instructor login and just participate from the student’s perspective in the ‘Course Overview’. That counts as student build plan as well.
    - The build plan for the student should now exist in the Jenkins folder of the exercise.
    - Delete {the whole folder of the exercise, only one of the base/solution build plans, both the base and solution build plans at the same time} in Jenkins. See the screenshots below on how to delete the folder/only single build plans.
    - Run the integrity check in Artemis again. The missing build plans of the base and/or solution build plans should be reported accordingly. The consistency check does not look at student build plans. Therefore, they are not reported as missing if you delete the whole folder.
    - Use the ‘Recreate Build Plans’ button in Artemis.
    - Now in Jenkins the folder and the base/solution build plans should be present again. This should also be reported by the integrity check. The student build plan has not been removed unless you deleted the whole folder.
    - Submit again from the student perspective. The build plan was recreated and you received build results.
        - If you deleted the whole folder, the student build plan cannot be recreated automatically. You have to run the ‘Cleanup’ action from the instructor view of the exercise first.
        - This is intended. Artemis does not know about the data-loss in Jenkins and assumes a build plan already exists, so it does not try to recreate it (checking for existence on every code push for every student would be too expensive). The instructor has to manually trigger the ‘Cleanup’ to force a recreation of student build plans on the next submit.

<details>
<summary>Screenshots of the relevant parts in the Jenkins UI</summary>

Deleting the whole folder on this page (:no_entry_sign: in the sidebar) completely removes the exercise from Jenkins including template, solution, and all student build plans.
![Screenshot 2022-02-23 at 13-08-50 All  GWTCBUILDPLANRECREATE   Jenkins](https://user-images.githubusercontent.com/64250573/155318109-38648494-0746-4005-b36c-4f80af8581dd.png)

By clicking on the SOLUTION/BASE-build plans you get to this page. There you can delete only this specific build plan (:no_entry_sign: to the left), the remaining build plans of the exercise remain.
![Screenshot 2022-02-23 at 13-10-18 GWTCBUILDPLANRECREATE-SOLUTION  GWTCBUILDPLANRECREATE   Jenkins](https://user-images.githubusercontent.com/64250573/155318117-5bf28265-045b-4a97-9fab-216908c9b854.png)
</details>

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage
| File | Branch | Statement |
| --- | --- | --- |
| `JenkinsService` | 88% | 94% |
| `JenkinsBuildPlanService` | 86% | 95% |

### Screenshots
The recreate build plan now received a confirmation dialogue explaining the consequences in more detail:
![Screenshot 2022-03-01 at 09-19-39 Programming Exercises](https://user-images.githubusercontent.com/64250573/156131555-2d631316-a193-40b5-9b8a-7c13d39de439.png)

The cleanup dialogue message has been extended to let instructors know that individual student build plans can be recreated automatically when the student submits again:
![Screenshot 2022-03-01 at 09-34-21 Programming Exercises](https://user-images.githubusercontent.com/64250573/156133756-c726653c-af5a-4d12-9f01-ead39f04d175.png)
